File: app/src/main/java/com/bopr/android/smailer/GoogleMail.java
package com.bopr.android.smailer;

import android.accounts.AccountsException;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.gmail.Gmail;
import com.google.api.services.gmail.model.ListMessagesResponse;
import com.google.api.services.gmail.model.Message;
import com.google.api.services.gmail.model.MessagePart;
import com.google.api.services.gmail.model.MessagePartHeader;
import com.google.api.services.gmail.model.ModifyMessageRequest;
import com.google.common.collect.ImmutableList;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.Properties;

import javax.activation.DataHandler;
import javax.activation.FileDataSource;
import javax.mail.Address;
import javax.mail.MessagingException;
import javax.mail.Multipart;
import javax.mail.Session;
import javax.mail.internet.AddressException;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeBodyPart;
import javax.mail.internet.MimeMessage;
import javax.mail.internet.MimeMultipart;

import static com.bopr.android.smailer.util.Util.isEmpty;
import static com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential.usingOAuth2;
import static com.google.api.client.util.Base64.decodeBase64;
import static com.google.api.client.util.StringUtils.newStringUtf8;
import static java.util.Arrays.asList;
import static javax.mail.Message.RecipientType.TO;

/**
 * Gmail mail transport.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class GoogleMail {

    private static Logger log = LoggerFactory.getLogger("GoogleMail");

    private static final String ME = "me"; /* exact lowercase "me" */
    private static final String UTF_8 = "UTF-8";
    private static final String HTML = "html";

    private final Context context;
    private Session session;
    private Gmail service;
    private String sender;

    public GoogleMail(@NonNull Context context) {
        this.context = context;
    }

    public void startSession(@NonNull String accoun, @NonNull String... scopes) throws AccountsException {
        this.sender = accoun;
        service = createService(createCredential(accoun, scopes));
        session = Session.getDefaultInstance(new Properties(), null);
    }

    public void send(@NonNull MailMessage message) throws IOException {
        service.users()
                .messages()
                .send(ME, createContent(message))
                .execute();

        log.debug("Message sent");
    }

    public List<MailMessage> list(@NonNull String query) throws IOException {
        ListMessagesResponse response = service
                .users()
                .messages()
                .list(ME)
                .setQ(query)
                .execute();

        LinkedList<MailMessage> result = new LinkedList<>();
        List<Message> messages = response.getMessages();
        if (messages != null) {
            for (Message m : messages) {
                Message message = service
                        .users()
                        .messages()
                        .get(ME, m.getId())
                        .execute();

                result.add(readMessage(message));
            }
        }
        return result;
    }

    void markAsRead(@NonNull MailMessage message) throws IOException {
        ModifyMessageRequest content = new ModifyMessageRequest()
                .setRemoveLabelIds(ImmutableList.of("UNREAD")); /* case sensitive */
        service.users()
                .messages()
                .modify(ME, message.getId(), content)
                .execute();

        log.debug("Message marked as read: " + message.getId());
    }

    void trash(@NonNull MailMessage message) throws IOException {
        service.users()
                .messages()
                .trash(ME, message.getId())
                .execute();

        log.debug("Message moved to trash: " + message.getId());
    }

    @NonNull
    private GoogleAccountCredential createCredential(String accountName, String... scopes) throws AccountsException {
        GoogleAccountCredential credential = usingOAuth2(context, asList(scopes));
        credential.setSelectedAccountName(accountName);
        if (credential.getSelectedAccount() == null) {
            throw new AccountsException("Account does not exist: " + accountName);
        }
        return credential;
    }

    @NonNull
    private Gmail createService(@NonNull GoogleAccountCredential credential) {
        return new Gmail.Builder(new NetHttpTransport(),
                JacksonFactory.getDefaultInstance(), credential)
                .setApplicationName("smailer")
                .build();
    }

    @NonNull
    private Message createContent(MailMessage message) {
        try {
            MimeMessage mimeMessage = new MimeMessage(session);
            mimeMessage.setFrom(sender);
            mimeMessage.setSubject(message.getSubject(), UTF_8);
            mimeMessage.setRecipients(TO, parseAddresses(message.getRecipients()));

            if (!isEmpty(message.getReplyTo())) {
                mimeMessage.setReplyTo(parseAddresses(message.getReplyTo()));
            }

            if (message.getAttachment() == null) {
                mimeMessage.setText(message.getBody(), UTF_8, HTML);
            } else {
                mimeMessage.setContent(createMultipart(message.getBody(), message.getAttachment()));
            }

            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
            mimeMessage.writeTo(buffer);
            return new Message().encodeRaw(buffer.toByteArray());

        } catch (IOException | MessagingException x) {
            throw new RuntimeException("Message creation failed", x);
        }
    }

    @NonNull
    private Multipart createMultipart(String body, @NonNull Collection<File> attachment) throws MessagingException {
        Multipart content = new MimeMultipart();

        MimeBodyPart textPart = new MimeBodyPart();
        textPart.setText(body, UTF_8, HTML);
        content.addBodyPart(textPart);

        for (File file : attachment) {
            MimeBodyPart attachmentPart = new MimeBodyPart();
            attachmentPart.setFileName(file.getName());
            attachmentPart.setDataHandler(new DataHandler(new FileDataSource(file)));
            content.addBodyPart(attachmentPart);
        }

        return content;
    }

    private Address[] parseAddresses(@NonNull String addresses) throws AddressException {
        if (addresses.indexOf(',') > 0) {
            return InternetAddress.parse(addresses);
        } else {
            return new InternetAddress[]{new InternetAddress(addresses)};
        }
    }

/*
    private MimeMessage toMimeMessage(Message message) throws MessagingException {
        String raw = newStringUtf8(decodeBase64(message.getRaw()));
        return new MimeMessage(session, new ByteArrayInputStream(raw.getBytes()));
    }
*/

    private MailMessage readMessage(Message gmailMessage) {
        MailMessage message = new MailMessage();
        message.setId(gmailMessage.getId());
        message.setSubject(readHeader(gmailMessage, "subject"));
        message.setFrom(readHeader(gmailMessage, "from"));
        message.setBody(readBody(gmailMessage));
        return message;
    }

    @Nullable
    private String readHeader(Message message, String name) {
        for (MessagePartHeader header : message.getPayload().getHeaders()) {
            if (header.getName().equalsIgnoreCase(name)) {
                return header.getValue();
            }
        }
        return null;
    }

    @Nullable
    private String readBody(Message message) {
        MessagePart payload = message.getPayload();
        if (payload != null) {
            List<MessagePart> parts = payload.getParts();
            if (parts == null) {
                return newStringUtf8(decodeBase64(payload.getBody().getData()));
            } else if (!parts.isEmpty()) {
                return newStringUtf8(decodeBase64(parts.get(0).getBody().getData()));
            }
        }
        return null;
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/GoogleMail.kt
package com.bopr.android.smailer

import android.accounts.Account
import android.content.Context
import com.bopr.android.smailer.util.Mockable
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.api.client.util.Base64
import com.google.api.client.util.StringUtils
import com.google.api.services.gmail.Gmail
import com.google.api.services.gmail.model.Message
import com.google.api.services.gmail.model.ModifyMessageRequest
import com.google.common.collect.ImmutableList
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.util.*
import javax.activation.DataHandler
import javax.activation.FileDataSource
import javax.mail.Message.RecipientType.TO
import javax.mail.MessagingException
import javax.mail.Multipart
import javax.mail.Session
import javax.mail.internet.*

/**
 * Gmail mail transport.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@Mockable
class GoogleMail(private val context: Context) {

    private val log = LoggerFactory.getLogger("GoogleMail")
    private lateinit var service: Gmail
    private lateinit var account: Account

    fun login(account: Account, scope: String) {
        this.account = account
        val credential = GoogleAccountCredential
                .usingOAuth2(context, listOf(scope))
                .setSelectedAccount(account)
        service = Gmail.Builder(NetHttpTransport(), JacksonFactory.getDefaultInstance(), credential)
                .setApplicationName("smailer")
                .build()
    }

    @Throws(IOException::class)
    fun send(message: MailMessage) {
        service.users()
                .messages()
                .send(ME, createContent(message))
                .execute()

        log.debug("Message sent")
    }

    @Throws(IOException::class)
    fun list(query: String): List<MailMessage> {
        val response = service
                .users()
                .messages()
                .list(ME)
                .setQ(query)
                .execute()

        val result = LinkedList<MailMessage>()
        response.messages?.let {
            for (m in it) {
                val message = service
                        .users()
                        .messages()[ME, m.id]
                        .execute()
                result.add(readMessage(message))
            }
        }
        return result
    }

    @Throws(IOException::class)
    fun markAsRead(message: MailMessage) {
        val content = ModifyMessageRequest()
                .setRemoveLabelIds(ImmutableList.of("UNREAD")) /* case sensitive */
        service.users()
                .messages()
                .modify(ME, message.id, content)
                .execute()

        log.debug("Message marked as read: " + message.id)
    }

    @Throws(IOException::class)
    fun trash(message: MailMessage) {
        service.users()
                .messages()
                .trash(ME, message.id)
                .execute()

        log.debug("Message moved to trash: " + message.id)
    }

    private fun createContent(message: MailMessage): Message {
        try {
            val session = Session.getDefaultInstance(Properties(), null)
            val mime = MimeMessage(session).apply {
                setFrom(message.from)
                setSubject(message.subject, UTF_8)
                setRecipients(TO, parseAddresses(message.recipients!!))
                if (!message.replyTo.isNullOrEmpty()) {
                    replyTo = parseAddresses(message.replyTo)
                }
                if (!message.attachment.isNullOrEmpty()) {
                    setContent(createMultipart(message.body, message.attachment))
                } else {
                    setText(message.body, UTF_8, HTML)
                }
            }
            
            val buffer = ByteArrayOutputStream()
            mime.writeTo(buffer)
            return Message().encodeRaw(buffer.toByteArray())

        } catch (x: IOException) {
            throw RuntimeException("Message creation failed", x)
        } catch (x: MessagingException) {
            throw RuntimeException("Message creation failed", x)
        }
    }

    @Throws(MessagingException::class)
    private fun createMultipart(body: String?, attachment: Collection<File>): Multipart {
        val content: Multipart = MimeMultipart()

        val textPart = MimeBodyPart().apply {
            setText(body, UTF_8, HTML)
        }
        content.addBodyPart(textPart)

        for (file in attachment) {
            val filePart = MimeBodyPart().apply {
                fileName = file.name
                dataHandler = DataHandler(FileDataSource(file))
            }
            content.addBodyPart(filePart)
        }

        return content
    }

    @Throws(AddressException::class)
    private fun parseAddresses(addresses: String): Array<out InternetAddress> {
        return if (addresses.indexOf(',') > 0) {
            InternetAddress.parse(addresses)
        } else {
            arrayOf(InternetAddress(addresses))
        }
    }

    private fun readMessage(gmailMessage: Message): MailMessage {
        return MailMessage(
                id = gmailMessage.id,
                subject = readHeader(gmailMessage, "subject"),
                from = readHeader(gmailMessage, "from"),
                body = readBody(gmailMessage)
        )
    }

    private fun readHeader(message: Message, name: String): String? {
        for (header in message.payload.headers) {
            if (header.name.equals(name, true)) {
                return header.value
            }
        }
        return null
    }

    private fun readBody(message: Message): String? {
        return message.payload?.let {
            val parts = it.parts
            val part = if (parts == null || parts.isEmpty()) it else parts[0]
            return StringUtils.newStringUtf8(Base64.decodeBase64(part.body.data))
        }
    }

    companion object {

        private const val ME = "me" /* exact lowercase "me" */
        private const val UTF_8 = "UTF-8"
        private const val HTML = "html"
    }

}